### PR TITLE
Adding an interpreter to lcalc

### PR DESCRIPTION
**Pull request by adelaett**
work in progress: I have some issues with GADT

### CODE REVIEW

File: compiler/shared_ast/interpreter.ml [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+            Errors.raise_spanned_error (Expr.mark_pos m) "Assertion failed")
+        | _ ->
+          Errors.raise_spanned_error (Expr.pos e')
+            "Expected a boolean literal for the result of this assertion \
+             (should not happen if the term was well-typed)")
+
+(** {1 API} *)
+
+let interpret_program (ctx : decl_ctx) (e : ('t, 'm) gexpr) :
+    (Uid.MarkedString.info * ('t, 'm) gexpr) list =
```
Comment (adelaett) [inline compiler/shared_ast/interpreter.ml:555]: This function should take a program as input instead.

**Review by adelaett — COMMENTED** [review]
Thanks @AltGr for this contribution. Is it possible to rework a little bit the cli frontend so we can do `catala interpret --lcalc blabla.catala_en` or `catala interpret --dcalc blabla.catala_en` ?

**Comment by AltGr** [comment]
> Thanks @AltGr for this contribution. Is it possible to rework a little bit the cli frontend so we can do `catala interpret --lcalc blabla.catala_en` or `catala interpret --dcalc blabla.catala_en` ?

It would probably be simpler to just add a toplevel `interpret-lcalc` command ?

File: compiler/shared_ast/interpreter.mli [hunk 2]
```diff
@@ -17,13 +17,16 @@
...
-val evaluate_expr : decl_ctx -> 'm Ast.expr -> 'm Ast.expr
+val evaluate_expr :
+  decl_ctx -> (([< dcalc | lcalc ] as 'a), 'm mark) gexpr -> ('a, 'm mark) gexpr
 (** Evaluates an expression according to the semantics of the default calculus. *)
 
 val interpret_program :
-  decl_ctx -> 'm Ast.expr -> (Uid.MarkedString.info * 'm Ast.expr) list
+  decl_ctx ->
+  (dcalc, 'm mark) gexpr ->
+  (Uid.MarkedString.info * (dcalc, 'm mark) gexpr) list
```
[the remark this answers is not in the document]
Comment (AltGr) [reply compiler/shared_ast/interpreter.mli:29]: Hm I think it's a leftover in the `mli` only, the function should actually work for both

File: compiler/catala_utils/cli.ml [hunk 3]
```diff
@@ -58,6 +60,7 @@ let backend_option_to_string = function
 let backend_option_of_string backend =
   match String.lowercase_ascii backend with
   | "interpret" -> `Interpret
+  | "interpret_lcalc" -> `Interpret_Lcalc
```
Comment (denismerigoux) [inline compiler/catala_utils/cli.ml:63]: This new backend should be documented in the `info` variable in this file (around line ~300).

File: compiler/lcalc/compile_with_exceptions.ml [hunk 4]
```diff
@@ -73,12 +73,12 @@ and translate_expr (ctx : 'm ctx) (e : 'm D.expr) : 'm A.expr boxed =
       (Marked.get_mark e)
   | EDefault { excepts; just; cons } ->
     translate_default ctx excepts just cons (Marked.get_mark e)
-  | ( ELit _ | EApp _ | EOp _ | EArray _ | EVar _ | EAbs _ | EIfThenElse _
-    | ETuple _ | ETupleAccess _ | EInj _ | EAssert _ | EStruct _
-    | EStructAccess _ | EMatch _ ) as e ->
-    Expr.map_raw ~fop:Operator.translate
-      ~floc:(function _ -> .)
-      ~f:(translate_expr ctx) (Marked.mark m e)
+  | EOp { op; tys } -> Expr.eop (Operator.translate op) tys m
```
Comment (denismerigoux) [inline compiler/lcalc/compile_with_exceptions.ml:76]: You're changing the behavior of your compilation pass here, is this wanted?

File: compiler/shared_ast/interpreter.ml [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+              e2
+          | _ ->
+            Cli.debug_format "%a" (Expr.format ctx) e';
+            Errors.raise_spanned_error (Expr.mark_pos m) "Assertion failed")
+        | _ ->
+          Errors.raise_spanned_error (Expr.pos e')
+            "Expected a boolean literal for the result of this assertion \
+             (should not happen if the term was well-typed)")
+
+let interpret_program_lcalc p s : (Uid.MarkedString.info * ('a, 'm) gexpr) list
```
Comment (denismerigoux) [inline compiler/shared_ast/interpreter.ml:558]: Do you really have to duplicate the code for `interpret_program_lcalc` and `interpret_program_dcalc` ?

**Review by denismerigoux — APPROVED** [review]
Nice job @adelaett!

File: compiler/shared_ast/interpreter.ml (hunk 5 above)
Comment (adelaett) [reply compiler/shared_ast/interpreter.ml:558]: yes there is typing issues. From the discussion with @AltGr, we concluded we need two differents functions.

File: compiler/lcalc/compile_with_exceptions.ml (hunk 4 above)
Comment (adelaett) [reply compiler/lcalc/compile_with_exceptions.ml:76]: We now have a different Expr.map function that don't required anymore the fop and floc functions. Hence, the behavior is the same for the lines you underlined.

File: compiler/catala_utils/cli.ml (hunk 3 above)
Comment (adelaett) [reply compiler/catala_utils/cli.ml:63]: I've added a docstring. Can you confirm it is ok?

**Comment by denismerigoux** [comment]
Perfect, we can merge once CI is green on my side.

**Pull request merged**